File: packages/lua/priorityqueue.lua
local priorityqueue = {}
local heap = require("heap")

local PriorityQueue = {}
PriorityQueue.__index = PriorityQueue

function priorityqueue.NewMinPQ()
    local pq = {
        h = heap.New(function(a, b) return a.priority < b.priority end)
    }
    return setmetatable(pq, PriorityQueue)
end

function priorityqueue.NewMaxPQ()
    local pq = {
        h = heap.New(function(a, b) return a.priority > b.priority end)
    }
    return setmetatable(pq, PriorityQueue)
end

function PriorityQueue:Push(value, priority)
    self.h:Push({value = value, priority = priority})
end

function PriorityQueue:Pop()
    local item, ok = self.h:Pop()
    if not ok then return nil, false end
    return item.value, true
end

function PriorityQueue:Peek()
    local item, ok = self.h:Peek()
    if not ok then return nil, false end
    return item.value, true
end

function PriorityQueue:Size()
    return self.h:Size()
end

return priorityqueue

File: packages/rust/src/priorityqueue.rs
use std::collections::BinaryHeap;
use std::cmp::Ordering;

#[derive(Eq, PartialEq)]
struct Node<T: Eq + PartialEq> {
    priority: i32,
    item: T,
}

impl<T: Eq + PartialEq> Ord for Node<T> {
    fn cmp(&self, other: &Self) -> Ordering {
        self.priority.cmp(&other.priority)
    }
}

impl<T: Eq + PartialEq> PartialOrd for Node<T> {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

pub struct PriorityQueue<T: Eq + PartialEq> {
    heap: BinaryHeap<Node<T>>,
}

impl<T: Eq + PartialEq> PriorityQueue<T> {
    pub fn new() -> Self {
        PriorityQueue { heap: BinaryHeap::new() }
    }

    pub fn push(&mut self, item: T, priority: i32) {
        self.heap.push(Node { priority, item });
    }

    pub fn pop(&mut self) -> Option<T> {
        self.heap.pop().map(|node| node.item)
    }
}
